src/FSharpPlus/Operators.fs

```diff
@@ -345,7 +345,60 @@ module Operators =
     let inline maxBy (projection: 'T->'U when 'U : comparison) (source: '``Foldable<'T>``) = MaxBy.Invoke projection  source : 'T
     let inline minBy (projection: 'T->'U when 'U : comparison) (source: '``Foldable<'T>``) = MinBy.Invoke projection  source : 'T
     let inline nth (n: int) (source: '``Foldable<'T>``) : 'T = Nth.Invoke n source
-       
+
+    /// <summary>
+    /// Returns the index of the first element in the foldable
+    /// that satisfies the given predicate.
+    /// </summary>
+    /// <param name="predicate">
+    /// The function to test the input elements.
+    /// </param>
+    /// <param name="source">The input foldable.</param>
+    /// <returns> 
+    /// The index of the first element that satisfies the predicate.
+    /// </returns>
+    /// <exception cref="System.ArgumentException">
+    /// Thrown if the predicate evaluates to false for all the elements of the source.
+    /// </exception>
+    let inline findIndex (predicate: 'T -> bool) (source: '``Foldable<'T>``) : int = FindIndex.Invoke predicate source
```

Comment:
Maybe I'm thinking too much, but what if we decide later to add a `findIndex` (and/or the other functions in this PR) to a multidimensional array?

The return type won't be a single int anymore, it would be a pair of ints.

If we think about GetSlice it considers a slice a variable type, becuse of the multidimensional arrays.

Of course, this might complicate type inference.

Thoughts?

Reply:
multidimensionals are not foldable anymore, so we must invent a new abstraction to do it

Reply:
Yes, I forgot to mention that. Although we might consider them as Foldables, but in that case the return type will be always an int.

If we want to have a generic return type we need to move out of Foldable, maybe Collection?

Reply:
Collections seems fine to me. I'll do it

Reply:
Another alternative, is to leave this as it is, and eventually create another set of (more) generic functions, as it's already the case with `nth` and `item`.

Then we can use the plural `indices` instead of `index`, and we can consider the `Indexable` abstraction.

Reply:
Plural `indices` sounds more like `('a -> bool) -> Collection<'a> -> int []` and I'm personally against that. 
Does `foldi` works on multidimentionals? If so, it will make more sense to place `findIndex` functions to them.

Reply:
Good question. TBH I can't recall, I did the Indexable abstraction (actually I took it from somewhere) long time ago :)

Reply:
Looking at the code, it seems to me that `foldi` for multidimensionals could be added. I think I was lazy, normally that's always the case ;)

Reply:
I think we need to add generic `getSlice` (or `slice`) built on top of `Indexable`. (`item` - `findIndex ((=) index)`, `getSlice` - `findSliceIndex`)

Reply:
Sounds good in principle !

Reply:
Anyway, if making the functions having generic return type (rather than `int`) doesn't break something, I think it is good to do it for now. We can later move them to `Indexable` without much pain.

Reply:
Or is it better to leave it as is?

Reply:
They might break, as `int` drives type inference. That's why I think having a second set of the functions might be desirable, although you can argue that's overkill.

Reply:
Btw: I'm also thinking, apart from multidimensionals arrays, `findIndex` could work on Dictionaries. Not the findSlice families, as they require a notion of contiguity.
Sorry, if I'm thinking too much, it's not my intent to complicate this PR with all this. What you did makes totally sense, I'm just thinking how can we leave room for later improvements.

Reply:
Perhaps for ordered dictionary it makes sense

Reply:
I like to handle all the case with just a single set of functions since it is less confusing to me. My idea is to make their return type generic for now to prevent future breaking changes in type signature. Doint it seems not breaking the currently supported types (foldable, string, array, etc).

Reply:
Let's try to be pragmatic as I want to merge this I see 2 options:

 - We find and use another name for "index" which means and integer index and just rename these functions.
 - We remove the int annotation and move these functions to Indexable.

Reply:
For the first option, the term "indexer" is also used for the syntax accessing dictionaries so the meaning of "index" is not limited to integer. So I think we don't have to invent a new word.

I complete agree with the second option.

Reply:
I agree as well, @wallymathieu any thoughts? Ok for the 2nd?

Reply:
In that option 2 we would have to remove the Default1 implementation as it is biased to an `int` index coming from a fold operation. Personally I'm fine with it.

Reply:
I'm fine with it as well.